File: MiniDOMBase/DMD/Sistema/Calendar/ActivePerson/CActivePerson.cs
using System;
using System.Collections;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using static minidom.Sistema;
using static minidom.Anagrafica;
using minidom.repositories;

namespace minidom
{
    public partial class Sistema
    {


        /// <summary>
        /// Rappresenta una persona "attiva" cioè per cui è stata definita almeno una attività nel calendario
        /// </summary>
        /// <remarks></remarks>
        [Serializable]
        public class CActivePerson
            : IComparable, IDMDXMLSerializable, IComparable<CActivePerson>
        {
            private const int GIORNATA_INTERA = 1;    // Flag utilizzato nel campo m_Flags (proprietà GiornataIntera)
            private const int PERSONA_GIURIDICA = 2;     // Settato se si tratta di una persona giuridica

            // Private m_IDRicontatto As Integer   'ID dell'oggetto ricontatto
            [NonSerialized] private Anagrafica.CRicontatto m_Ricontatto; // Oggetto ricontatto
            private int m_PersonID;       // ID della persona da ricontattare
            [NonSerialized] private Anagrafica.CPersona m_Person;        // Persona da ricontattare
            private string m_Nominativo;      // Nome della persona o dell'azienda da ricontattare
            private CCollection m_Activities; // Azioni proposte
            private string m_Notes;           // Stringa descrittiva del ricontatto
            private DateTime? m_Data; // Data del ricontatto
            private int m_Flags;          // Flags
            private int m_Promemoria;     // Promemoria in minuti
            private CKeyCollection m_MoreInfo; // (Of String)
            private string m_Categoria;
            private string m_IconURL;

            /// <summary>
            /// Costruttore
            /// </summary>
            public CActivePerson()
            {
               // //DMDObject.IncreaseCounter(this);
                // Me.m_IDRicontatto = 0
                m_Ricontatto = null;
                m_PersonID = 0;
                m_Person = null;
                m_Nominativo = "";
                m_Activities = null;
                m_Notes = "";
                m_Data = default;
                m_Flags = 0;
                m_Promemoria = 0;
                m_MoreInfo = new CKeyCollection(); // (Of Object)
                m_Categoria = "";
                m_IconURL = "";
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="ric"></param>
            public CActivePerson(Anagrafica.CRicontatto ric) : this()
            {
                if (ric is null)
                    throw new ArgumentNullException("ric");
                Ricontatto = ric;
                Data = ric.DataPrevista;
                Notes = ric.Note;
                GiornataIntera = ric.GiornataIntera;
                Promemoria = ric.Promemoria;
                Categoria = ric.Categoria;
                PersonID = ric.IDPersona;
                Person = ric.Persona;
                if (Person is object)
                {
                    IconURL = Person.IconURL;
                }
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="p"></param>
            public CActivePerson(Anagrafica.CPersona p) : this()
            {
                if (p is null)
                    throw new ArgumentNullException("p");
                Ricontatto = new Anagrafica.CRicontatto();
                Ricontatto.Persona = p;
                Ricontatto.DataPrevista = DMD.DateUtils.Now();
                Ricontatto.StatoRicontatto = (Anagrafica.StatoRicontatto)1;
                Data = Ricontatto.DataPrevista;
                Notes = "";
                GiornataIntera = true;
                Promemoria = 0;
                Categoria = "Normale";
                PersonID = DBUtils.GetID(p, 0);
                Person = p;
                IconURL = Person.IconURL;
            }

            void IDMDXMLSerializable.XMLSerialize(XMLWriter writer)
            {
                this.XMLSerialize(writer);
            }

            void IDMDXMLSerializable.SetFieldInternal(string fieldName, object fieldValue)
            {
                this.SetFieldInternal(fieldName, fieldValue);
            }

            /// <summary>
            /// Restituisce o imposta la categoria del ricontatto
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string Categoria
            {
                get
                {
                    return m_Categoria;
                }

                set
                {
                    m_Categoria = DMD.Strings.Trim(value);
                }
            }

            /// <summary>
            /// Restituisce o imposta l'icona associata al contatto
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string IconURL
            {
                get
                {
                    return m_IconURL;
                }

                set
                {
                    m_IconURL = DMD.Strings.Trim(value);
                }
            }

            /// <summary>
            /// Vero se si tratta di un evento per cui non è stato fissato un orario particolare
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public bool GiornataIntera
            {
                get
                {
                    return DMD.RunTime.TestFlag(m_Flags, GIORNATA_INTERA);
                }

                set
                {
                    m_Flags = DMD.RunTime.SetFlag(m_Flags, GIORNATA_INTERA, value);
                }
            }

            /// <summary>
            /// Vero se la persona è una persona giuridica
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public bool PersonaGiuridica
            {
                get
                {
                    return DMD.RunTime.TestFlag(m_Flags, PERSONA_GIURIDICA);
                }

                set
                {
                    m_Flags = DMD.RunTime.SetFlag(m_Flags, PERSONA_GIURIDICA, value);
                }
            }

            // ''' <summary>
            // ''' ID del ricontatto corrispondente
            // ''' </summary>
            // ''' <value></value>
            // ''' <returns></returns>
            // ''' <remarks></remarks>
            // Public Property IDRicontatto As Integer
            // Get
            // Return GetID(Me.m_Ricontatto, Me.m_IDRicontatto)
            // End Get
            // Set(value As Integer)
            // If Me.IDRicontatto = value Then Exit Property
            // Me.m_IDRicontatto = value
            // Me.m_Ricontatto = Nothing
            // End Set
            // End Property

            /// <summary>
            /// Ricontatto corrispondente
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public Anagrafica.CRicontatto Ricontatto
            {
                get
                {
                    // If Me.m_Ricontatto Is Nothing Then Me.m_Ricontatto = Ricontatti.GetItemById(Me.m_IDRicontatto)
                    return m_Ricontatto;
                }

                set
                {
                    m_Ricontatto = value;
                    // Me.m_IDRicontatto = GetID(value)
                }
            }

            /// <summary>
            /// ID della persona
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public int PersonID
            {
                get
                {
                    return DBUtils.GetID(m_Person, m_PersonID);
                }

                set
                {
                    if (PersonID == value)
                        return;
                    m_PersonID = value;
                    m_Person = null;
                }
            }

            /// <summary>
            /// Persona
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public Anagrafica.CPersona Person
            {
                get
                {
                    if (m_Person is null)
                        m_Person = Anagrafica.Persone.GetItemById(m_PersonID);
                    return m_Person;
                }

                set
                {
                    m_Person = value;
                    m_PersonID = DBUtils.GetID(value, 0);
                    if (value is object)
                    {
                        m_Nominativo = value.Nominativo;
                        PersonaGiuridica = value.TipoPersona != Anagrafica.TipoPersona.PERSONA_FISICA;
                    }
                }
            }

            /// <summary>
            /// Nominativo della persona
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string Nominativo
            {
                get
                {
                    return m_Nominativo;
                }

                set
                {
                    m_Nominativo = DMD.Strings.Trim(value);
                }
            }

            /// <summary>
            /// Azioni possibili sul ricontatto
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public CCollection Activities
            {
                get
                {
                    if (m_Activities is null)
                        m_Activities = new CCollection();
                    return m_Activities;
                }
            }

            /// <summary>
            /// Data per cui è stato fissato il ricontatto
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public DateTime? Data
            {
                get
                {
                    return m_Data;
                }

                set
                {
                    m_Data = value;
                }
            }

            /// <summary>
            /// Informazioni sul ricontatto
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string Notes
            {
                get
                {
                    return m_Notes;
                }

                set
                {
                    m_Notes = DMD.Strings.Trim(value);
                }
            }

            /// <summary>
            /// Promemoria in minuti
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public int Promemoria
            {
                get
                {
                    return m_Promemoria;
                }

                set
                {
                    m_Promemoria = value;
                }
            }

            /// <summary>
            /// Eventuali informazioni aggiuntive
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public CKeyCollection MoreInfo // (Of String)
            {
                get
                {
                    return m_MoreInfo;
                }
            }

            /// <summary>
            /// Serializzazione XML
            /// </summary>
            /// <param name="writer"></param>
            protected void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("PersonID", PersonID);
                writer.WriteAttribute("Nominativo", m_Nominativo);
                writer.WriteAttribute("Flags", m_Flags);
                writer.WriteAttribute("Categoria", m_Categoria);
                writer.WriteAttribute("IconURL", m_IconURL);
                writer.WriteAttribute("Promemoria", m_Promemoria);
                writer.WriteAttribute("Data", m_Data);
                writer.WriteTag("Ricontatto", Ricontatto);
                writer.WriteTag("Activities", Activities);
                writer.WriteTag("Notes", m_Notes);
                writer.WriteTag("MoreInfo", m_MoreInfo);
            }

            /// <summary>
            /// Deserializzazione xml
            /// </summary>
            /// <param name="fieldName"></param>
            /// <param name="fieldValue"></param>
            protected void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "PersonID":
                        {
                            m_PersonID = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "Ricontatto":
                        {
                            m_Ricontatto = (Anagrafica.CRicontatto)fieldValue; // DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue)
                            break;
                        }

                    case "Nominativo":
                        {
                            m_Nominativo = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Activities":
                        {
                            m_Activities = (CCollection)fieldValue;
                            break;
                        }

                    case "Data":
                        {
                            m_Data = DMD.XML.Utils.Serializer.DeserializeDate(fieldValue);
                            break;
                        }

                    case "Notes":
                        {
                            m_Notes = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Flags":
                        {
                            m_Flags = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "Promemoria":
                        {
                            m_Promemoria = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "MoreInfo":
                        {
                            m_MoreInfo = (CKeyCollection)fieldValue;
                            break;
                        }

                    case "Categoria":
                        {
                            m_Categoria = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "IconURL":
                        {
                            m_IconURL = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }
                }
            }

            /// <summary>
            /// Compara due oggetti
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public int CompareTo(CActivePerson obj)
            {
                var d = DMD.DateUtils.Now();
                var d1 = (this.GiornataIntera)? DMD.DateUtils.GetDatePart(this.m_Data) : this.m_Data;
                var d2 = (obj.GiornataIntera) ? DMD.DateUtils.GetDatePart(obj.m_Data) : obj.m_Data;
                long? diff1 = DMD.Longs.Add(DMD.DateUtils.DateDiff("m", d1, d), this.Promemoria);
                long? diff2 = DMD.Longs.Add(DMD.DateUtils.DateDiff("m", d2, d), obj.Promemoria);
                int ret = DMD.Longs.Compare(diff1 , diff2);
                if (ret == 0) ret = DMD.Strings.Compare(m_Nominativo, obj.m_Nominativo, true);
                return ret;
                 
            }

            int IComparable.CompareTo(object obj)
            {
                return CompareTo((CActivePerson)obj);
            }

            /// <summary>
            /// Distruttore
            /// </summary>
            ~CActivePerson()
            {
                ////DMDObject.DecreaseCounter(this);
            }
        }
    }
}

File: MiniDOMBase/DMD/Sistema/Calendar/CActivePerson.vb

Imports minidom
Imports minidom.Databases
Imports minidom.Sistema
Imports minidom.Anagrafica

Partial Class Sistema


    ''' <summary>
    ''' Rappresenta una persona "attiva" cioè per cui è stata definita almeno una attività nel calendario
    ''' </summary>
    ''' <remarks></remarks>
    Public Class CActivePerson
        Implements IComparable, XML.IDMDXMLSerializable

        Private Const GIORNATA_INTERA As Integer = 1    'Flag utilizzato nel campo m_Flags (proprietà GiornataIntera)
        Private Const PERSONA_GIURIDICA As Integer = 2     'Settato se si tratta di una persona giuridica

        'Private m_IDRicontatto As Integer   'ID dell'oggetto ricontatto
        Private m_Ricontatto As CRicontatto 'Oggetto ricontatto
        Private m_PersonID As Integer       'ID della persona da ricontattare
        Private m_Person As CPersona        'Persona da ricontattare
        Private m_Nominativo As String      'Nome della persona o dell'azienda da ricontattare
        Private m_Activities As CCollection 'Azioni proposte
        Private m_Notes As String           'Stringa descrittiva del ricontatto
        Private m_Data As Date? 'Data del ricontatto
        Private m_Flags As Integer          'Flags
        Private m_Promemoria As Integer     'Promemoria in minuti
        Private m_MoreInfo As CKeyCollection '(Of String)
        Private m_Categoria As String
        Private m_IconURL As String

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
            'Me.m_IDRicontatto = 0
            Me.m_Ricontatto = Nothing
            Me.m_PersonID = 0
            Me.m_Person = Nothing
            Me.m_Nominativo = ""
            Me.m_Activities = Nothing
            Me.m_Notes = ""
            Me.m_Data = Nothing
            Me.m_Flags = 0
            Me.m_Promemoria = 0
            Me.m_MoreInfo = New CKeyCollection '(Of Object)
            Me.m_Categoria = ""
            Me.m_IconURL = ""
        End Sub

        Public Sub New(ByVal ric As CRicontatto)
            Me.New
            If (ric Is Nothing) Then Throw New ArgumentNullException("ric")
            Me.Ricontatto = ric
            Me.Data = ric.DataPrevista
            Me.Notes = ric.Note
            Me.GiornataIntera = ric.GiornataIntera
            Me.Promemoria = ric.Promemoria
            Me.Categoria = ric.Categoria
            Me.PersonID = ric.IDPersona
            Me.Person = ric.Persona
            If (Me.Person IsNot Nothing) Then
                Me.IconURL = Me.Person.IconURL
            End If
        End Sub

        Public Sub New(ByVal p As CPersona)
            Me.New
            If (p Is Nothing) Then Throw New ArgumentNullException("p")
            Me.Ricontatto = New CRicontatto
            Me.Ricontatto.Persona = p
            Me.Ricontatto.DataPrevista = DateUtils.Now
            Me.Ricontatto.StatoRicontatto = 1
            Me.Data = Me.Ricontatto.DataPrevista
            Me.Notes = ""
            Me.GiornataIntera = True
            Me.Promemoria = 0
            Me.Categoria = "Normale"
            Me.PersonID = GetID(p)
            Me.Person = p
            Me.IconURL = Me.Person.IconURL
        End Sub

        ''' <summary>
        ''' Restituisce o imposta la categoria del ricontatto
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Categoria As String
            Get
                Return Me.m_Categoria
            End Get
            Set(value As String)
                Me.m_Categoria = Trim(value)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta l'icona associata al contatto
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property IconURL As String
            Get
                Return Me.m_IconURL
            End Get
            Set(value As String)
                Me.m_IconURL = Trim(value)
            End Set
        End Property

        ''' <summary>
        ''' Vero se si tratta di un evento per cui non è stato fissato un orario particolare
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property GiornataIntera As Boolean
            Get
                Return TestFlag(Me.m_Flags, GIORNATA_INTERA)
            End Get
            Set(value As Boolean)
                Me.m_Flags = SetFlag(Me.m_Flags, GIORNATA_INTERA, value)
            End Set
        End Property

        ''' <summary>
        ''' Vero se la persona è una persona giuridica
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property PersonaGiuridica As Boolean
            Get
                Return TestFlag(Me.m_Flags, PERSONA_GIURIDICA)
            End Get
            Set(value As Boolean)
                Me.m_Flags = SetFlag(Me.m_Flags, PERSONA_GIURIDICA, value)
            End Set
        End Property

        ' ''' <summary>
        ' ''' ID del ricontatto corrispondente
        ' ''' </summary>
        ' ''' <value></value>
        ' ''' <returns></returns>
        ' ''' <remarks></remarks>
        'Public Property IDRicontatto As Integer
        '    Get
        '        Return GetID(Me.m_Ricontatto, Me.m_IDRicontatto)
        '    End Get
        '    Set(value As Integer)
        '        If Me.IDRicontatto = value Then Exit Property
        '        Me.m_IDRicontatto = value
        '        Me.m_Ricontatto = Nothing
        '    End Set
        'End Property

        ''' <summary>
        ''' Ricontatto corrispondente
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Ricontatto As CRicontatto
            Get
                'If Me.m_Ricontatto Is Nothing Then Me.m_Ricontatto = Ricontatti.GetItemById(Me.m_IDRicontatto)
                Return Me.m_Ricontatto
            End Get
            Set(value As CRicontatto)
                Me.m_Ricontatto = value
                'Me.m_IDRicontatto = GetID(value)
            End Set
        End Property

        ''' <summary>
        ''' ID della persona
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property PersonID As Integer
            Get
                Return GetID(Me.m_Person, Me.m_PersonID)
            End Get
            Set(value As Integer)
                If (Me.PersonID = value) Then Exit Property
                Me.m_PersonID = value
                Me.m_Person = Nothing
            End Set
        End Property

        ''' <summary>
        ''' Persona
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Person As CPersona
            Get
                If (Me.m_Person Is Nothing) Then Me.m_Person = Anagrafica.Persone.GetItemById(Me.m_PersonID)
                Return Me.m_Person
            End Get
            Set(value As CPersona)
                Me.m_Person = value
                Me.m_PersonID = GetID(value)
                If Not (value Is Nothing) Then
                    Me.m_Nominativo = value.Nominativo
                    Me.PersonaGiuridica = value.TipoPersona <> TipoPersona.PERSONA_FISICA
                End If
            End Set
        End Property

        ''' <summary>
        ''' Nominativo della persona
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Nominativo As String
            Get
                Return Me.m_Nominativo
            End Get
            Set(value As String)
                Me.m_Nominativo = Trim(value)
            End Set
        End Property

        ''' <summary>
        ''' Azioni possibili sul ricontatto
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property Activities As CCollection
            Get
                If (Me.m_Activities Is Nothing) Then Me.m_Activities = New CCollection
                Return Me.m_Activities
            End Get
        End Property

        ''' <summary>
        ''' Data per cui è stato fissato il ricontatto
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Data As Date?
            Get
                Return Me.m_Data
            End Get
            Set(value As Date?)
                Me.m_Data = value
            End Set
        End Property

        ''' <summary>
        ''' Informazioni sul ricontatto
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Notes As String
            Get
                Return Me.m_Notes
            End Get
            Set(value As String)
                Me.m_Notes = Trim(value)
            End Set
        End Property

        ''' <summary>
        ''' Promemoria in minuti
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Promemoria As Integer
            Get
                Return Me.m_Promemoria
            End Get
            Set(value As Integer)
                Me.m_Promemoria = value
            End Set
        End Property

        ''' <summary>
        ''' Eventuali informazioni aggiuntive
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property MoreInfo As CKeyCollection '(Of String)
            Get
                Return Me.m_MoreInfo
            End Get
        End Property

        Protected Sub XMLSerialize(ByVal writer As minidom.XML.XMLWriter) Implements XML.IDMDXMLSerializable.XMLSerialize
            writer.WriteAttribute("PersonID", Me.PersonID)
            writer.WriteAttribute("Nominativo", Me.m_Nominativo)
            writer.WriteAttribute("Flags", Me.m_Flags)
            writer.WriteAttribute("Categoria", Me.m_Categoria)
            writer.WriteAttribute("IconURL", Me.m_IconURL)
            writer.WriteAttribute("Promemoria", Me.m_Promemoria)
            writer.WriteAttribute("Data", Me.m_Data)
            writer.WriteTag("Ricontatto", Me.Ricontatto)
            writer.WriteTag("Activities", Me.Activities)
            writer.WriteTag("Notes", Me.m_Notes)
            writer.WriteTag("MoreInfo", Me.m_MoreInfo)
        End Sub

        Protected Sub SetFieldInternal(ByVal fieldName As String, ByVal fieldValue As Object) Implements XML.IDMDXMLSerializable.SetFieldInternal
            Select Case fieldName
                Case "PersonID" : Me.m_PersonID = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "Ricontatto" : Me.m_Ricontatto = fieldValue 'XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "Nominativo" : Me.m_Nominativo = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Activities" : Me.m_Activities = fieldValue
                Case "Data" : Me.m_Data = XML.Utils.Serializer.DeserializeDate(fieldValue)
                Case "Notes" : Me.m_Notes = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Flags" : Me.m_Flags = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "Promemoria" : Me.m_Promemoria = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "MoreInfo" : Me.m_MoreInfo = fieldValue
                Case "Categoria" : Me.m_Categoria = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "IconURL" : Me.m_IconURL = XML.Utils.Serializer.DeserializeString(fieldValue)
            End Select
        End Sub

        Public Function CompareTo(ByVal obj As CActivePerson) As Integer
            If (Types.IsNull(Me.m_Data)) Then
                If (Types.IsNull(obj.m_Data)) Then
                    Return 0
                Else
                    Return -1
                End If
            Else
                If (Types.IsNull(obj.m_Data)) Then
                    Return 1
                Else
                    Dim d1 As Date = Me.m_Data
                    Dim d2 As Date = obj.m_Data
                    If (Me.GiornataIntera) Then d1 = DateUtils.GetDatePart(d1)
                    If (obj.GiornataIntera) Then d2 = DateUtils.GetDatePart(d2)
                    'Calcoliamo la distanza (in minuti) tra le due date e l'ora attuale sommandola ai relativi promemoria
                    Dim diff1 As Integer = DateUtils.DateDiff("m", d1, DateUtils.Now) + Me.Promemoria
                    Dim diff2 As Integer = DateUtils.DateDiff("m", d2, DateUtils.Now) + obj.Promemoria
                    Dim ret As Integer = diff1 - diff2
                    If (ret = 0) Then ret = Strings.Compare(Me.m_Nominativo, obj.m_Nominativo, CompareMethod.Text)
                    Return ret
                End If
            End If
        End Function

        Private Function CompareTo1(ByVal obj As Object) As Integer Implements IComparable.CompareTo
            Return Me.CompareTo(obj)
        End Function

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub
    End Class


End Class